x/config/configurator.go

```diff
@@ -0,0 +1,406 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"reflect"
+
+	"go.uber.org/yarpc/internal/decode"
+
+	"gopkg.in/yaml.v2"
+)
+
+// Configurator helps build Dispatchers using runtime configuration.
+type Configurator struct {
+	knownTransports map[string]transportSpec
+}
+
+// New sets up a new empty Configurator. The returned Configurator does not
+// know about any transports. Individual TransportSpecs must be registered
+// against it using the RegisterTransport function.
+func New() *Configurator {
+	return &Configurator{knownTransports: make(map[string]transportSpec)}
+}
+
+// TransportSpec specifies the configuration parameters for a transport. These
+// specifications are registered against a Configurator to teach it how to
+// parse the configuration for that transport and build instances of it.
+type TransportSpec struct {
+	// Name of the transport
+	Name string
+
+	// References to a TransportBuilder which specifies the configuration for
+	// a Transport. Transport instances are shared across all inbounds and
+	// outbounds of a specific transport type.
+	TransportConfigType TransportBuilder
+
+	// TODO(abg): Document how these values are actually used since it may be
+	// non-obvious.
+
+	// Everything below is optional
+
+	InboundConfigType        InboundBuilder
+	UnaryOutboundConfigType  UnaryOutboundBuilder
+	OnewayOutboundConfigType OnewayOutboundBuilder
+
+	UnaryOutboundPresets  map[string]UnaryOutboundBuilder
+	OnewayOutboundPresets map[string]OnewayOutboundBuilder
+}
+
+// RegisterTransport registers a TransportSpec with the given Configurator.
+func (c *Configurator) RegisterTransport(t TransportSpec) error {
+	getStruct := func(t reflect.Type) reflect.Type {
```

Comment:
this will reject &&struct am I correct? Error msgs later in this function mention: "only work on struct". But if you pass (likely by mistake) &&struct, the error might be confusing. "only work on struct and pointer to struct"?

Suggestion:

```go
getStruct := func(t reflect.Type, configtype string) (reflect.Type, err) {
	 switch t.Kind() {
 		case reflect.Struct:
 			return t, nil
 		case reflect.Ptr:
 			if t.Elem().Kind() == reflect.Struct {
 				return t.Elem(), nil
 			}
 		}
 		return nil, fmt.Errorf("%s configurations can only be defined on structs and pointer on structs", configtype)
}
```